change imviz reference data coords from GWCS to FITS WCS

bmorris3 opened the pull request:
### Background

Since #2179, Imviz displays images in the sky plane without distortion by creating a glue reference data layer that contains only coordinates (no image). The WCS-only layer was has its coordinates defined by a GWCS object. I chose GWCS over FITS WCS because it seemed to me at the time that encoding reference frame rotation in FITS WCS was nontrivial. Since then, I noticed that longitudinal pole rotation is supported by low level WCS.

### Why swap GWCS for FITS WCS in reference data?

1. The API for creating low level WCS is very very stable, whereas GWCS is evolving to support Roman. This lowers the risk of imviz maintenance burden on our end.
2. Low level WCS computes transformations in C, which will be as fast or faster than GWCS.

When we think about "linking performance", what we're usually talking about is limiting the number of links and removing redundant ones. The floor on the speed of calculations through links is set by the performance of the WCS transformations.

#### Profiling

I ran [profiling tests in this notebook](https://gist.github.com/bmorris3/3b63f6537b8649f2e50bd0b9edc25c3e) to do transformations using the coordinates from a simulated Roman WFI exposure from Build 16 (requires `roman_datamodels==0.22.0`). I compared the GWCS performance with FITS SIP computed by calling the `gwcs.to_fits_sip` method.

FITS WCS transformations are 1.5x faster than GWCS in pixel to world transformations computed on large grids (thousands of pixels per side). FITS WCS and GWCS have roughly equivalent performance in world to pixel transformations on large grids. When computing transformations on smaller numbers of pixels (10s-100s), FITS WCS outperforms GWCS by ~10x in both pixel to world and world to pixel transformations. The gap between FITS WCS and GWCS gets larger for smaller arrays due to overheads in GWCS, which internally initializes new objects on every transformation, so performance sees little improvement once arrays are smaller than ~100x100.

Aside: this profiling was performed on Roman observation coordinates with all of their complexity in both GWCS/FITS SIP representations. The reference data layer doesn't need any of that complexity – it's just the sky plane! The rotated WCS created for reference data in this PR is just RA---TAN/DEC--TAN ± rotation, without distortion.

![gwcs_fits_sip](https://github.com/user-attachments/assets/c4d56e03-a058-401c-bac0-26c1a1345412)

### Next steps

This PR swaps the reference data layer's WCS implementation from GWCS to FITS WCS. A second performance improvement will come from optionally approximating GWCS objects with FITS SIP when possible. When the SIP approximation can be made, this brings significant performance improvements for a small loss in accuracy.

The lower row of the profiling plot above shows the mean, standard deviation, and maximum of the difference between the GWCS and FITS SIP coordinates. The maximum difference is less about 0.1 pixels, where most FITS SIP coordinates are within 0.02 pixels of the GWCS coordinates. For reference, the astrometric precision of Roman source catalog entries from PSF fitting is expected to be about the same size (https://github.com/spacetelescope/romancal/pull/794).

Our goal is to let the user choose via API or UI whether or not to use this approximation for data that they've loaded, which is not enabled by this PR.

As part of this PR, I've added a hidden method `_try_gwcs_to_fits_sip` to the imviz parser which tries to approximate with FITS SIP for data that enters the parser with GWCS. This function never gets called in the code in this PR, but we can use it in the (near!) future when we add support for optional FITS SIP approx from GWCS input.


### Other notes

This PR removes a line in the imviz parser which removes the bounding box from GWCS on load. Among other changes, a similar change was suggested in #2887. That PR is not superseded by this one.

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

pllim commented on jdaviz/configs/imviz/plugins/coords_info/coords_info.py:
```diff
@@ -460,9 +461,15 @@ def _image_viewer_update(self, viewer, x, y):
 
         maxsize = int(np.ceil(np.log10(np.max(active_layer.layer.shape)))) + 3
-        fmt = 'x={0:0' + str(maxsize) + '.1f} y={1:0' + str(maxsize) + '.1f}'
+        if any(['nan' in map(str, (x, y))]):
```
I think NaN could also result from invalid FITS WCS conversion but I am not 100% sure now. No reason to remove unless somehow this also deteriorates the performance significantly?

bmorris3 replied:
I agree that this doesn't affect performance. 

On the other hand, is it meaningful or useful to display "nan" as a pixel coordinate?

pllim replied:
I think it just crashes if `np.nan` is not handled. I didn't add this logic for fun but the exact details, I can't quite recall now. I just know it was needed.

pllim commented on jdaviz/configs/imviz/plugins/parsers.py:
```diff
@@ -235,14 +235,6 @@ def _parse_image(app, file_obj, data_label, ext=None, parent=None):
-            # so that image layers are not cropped.
-            # NOTE: if extending this beyond GWCS, the mouseover logic
-            # for outside_*_bounding_box should also be updated.
-            data.coords._orig_bounding_box = data.coords.bounding_box
```
I do not understand why this is removed. There is still an option for user to choose GWCS. It was specifically requested that we ignore the bounding box because mouseover get messed up when there are bounding boxes and user looking at mosaic. This is not just for Orientation.

bmorris3 replied:
If (2) is acceptable, do you want to see that as part of this PR, or an immediate follow-up PR?

pllim replied:
I cannot speak for Roman use cases. Here is the original use case that required us to ignore the bounding box: User loads 4 JWST exposures (they have GWCS, not FITS WCS; but distortion is not an issue, just how parser grabbed WCS info). Let's forget about Orientation for a second, so first loaded image is the reference image. The exposures all get displayed at the same time because we support partial opacity. They have different bounding boxes. So when user mouse over to some parts outside one of them bounding boxes, it shows NaN. But user wants to see some real info because again, distortion is not problem here.

Roman support should not introduce regression to other non-Roman use cases. Or if it does, we have to ask stakeholders if they are okay with it.

pllim replied:
I think after offline discussions, we're in favor of keeping the changes as proposed, so I am marking this as resolved. Thanks for your time.

pllim commented on jdaviz/configs/imviz/plugins/parsers.py:
```diff
@@ -347,7 +339,27 @@ def _jwst_to_glue_data(file_obj, ext, data_label):
 
 
-def _jwst2data(file_obj, ext, data_label):
+def _try_gwcs_to_fits_sip(gwcs):
+    """
+    Try to convert this GWCS to FITS SIP. Some GWCS models
+    cannot be converted to FITS SIP. In that case, a warning
+    is raised and the GWCS is used, as is.
+    """
+    try:
+        result = WCS(gwcs.to_fits_sip())
+    except ValueError as err:
+        warnings.warn(
+            "The GWCS coordinates could not be simplified to "
+            "a SIP-based FITS WCS, the following error was "
+            f"raised: {err}",
+            UserWarning
+        )
+        result = gwcs
+
+    return result
```
Should this function be added when it is actually used then? If we put it in now and then forget about it and end up not using it, we will just carry dead code around.

bmorris3 replied:
As mentioned in the description, this will be used in the next PR, which I am working on next sprint. It won't be carried around long.

pllim commented on jdaviz/configs/imviz/plugins/parsers.py:
```diff
@@ -367,7 +379,14 @@ def _jwst2data(file_obj, ext, data_label):
+                gwcs = dm_meta['wcs']
+
+                if try_gwcs_to_fits_sip:
+                    coords = _try_gwcs_to_fits_sip(gwcs)
```
If you want FITS WCS anyway, can we just grab it from the FITS extension? The only reason I was accessing the ASDF-in-FITS was for the GWCS. By contract, JWST is required to also ship the FITS WCS in normal FITS header (even if it cannot represent all the distortion, or something like that), so maybe this conversion is not necessary.

bmorris3 replied:
Good point. I'll simplify this part of the parser to use FITS as FITS.

pllim commented on jdaviz/configs/imviz/tests/test_linking.py:
```diff
@@ -237,19 +237,21 @@ def test_wcslink_rotated(self):
-        assert label_mouseover.as_text() == ('Pixel x=-1.0 y=-1.0',
-                                             'World 00h14m19.5829s -30d23m30.9860s (ICRS)',
-                                             '3.5815955408 -30.3919405616 (deg)')
+        assert label_mouseover.as_text() == ('Pixel', '', '')
```
Shouldn't `'Pixel x=-1.0 y=-1.0'` still display? The mouse location is still valid, just the WCS is NaN.

bmorris3 replied:
I'm not sure. If we're self-consistently _not_ showing out-of-bounds coordinates, I think that should apply in both p2w/w2p transformations.

pllim replied:
I still want the X/Y even when out of bounds. Can you please put that back? Thanks.

pllim commented on jdaviz/configs/imviz/tests/test_linking.py:
```diff
@@ -281,26 +283,34 @@ def test_pixel_linking(self):
+        # `row2_unreliable` was True when the WCS-only layer used GWCS.
+        # now with FITS WCS as the coordinate frame, there is no bounding
+        # box and row2 is reliable.
+        assert not label_mouseover.row2_unreliable
```
What does "unreliable" even mean anymore now that we do not allow any extrapolation? Should this whole "unreliable" definition be removed? Maybe @kecnry can chime in too since I think he coded a lot of these.

kecnry replied:
If we _never_ set it to True anymore, then we can remove it.  If there are cases that use it, we might as well keep it.

pllim replied:
Where do we use it if we now follow GWCS bounding box and anything unreliable is always NaNs?

pllim commented:
In general I like the idea of using a simpler FITS WCS as the underlying "fake image" layer for Orientation. I am still curious about benchmark for (a) JWST, and (b) distorted HST/ACS exposures. These are accessible via Imviz example notebooks in this repo. Thanks!

camipacifici commented:
I was testing this to see if it solves the problem with the viewer that sometimes freezes and instead found that the align by WCS is broken for the dataset in this notebook: https://github.com/spacetelescope/jdat_notebooks/blob/main/notebooks/cross_instrument/photometry_extragalactic/imviz.ipynb
Can you take a look please? I see a traceback when asking to link by WCS from the API.

camipacifici commented:
I think this is broken already in main 🧐

bafio commented on jdaviz/configs/imviz/plugins/coords_info/coords_info.py:
```diff
@@ -460,9 +461,15 @@ def _image_viewer_update(self, viewer, x, y):
+            # don't show nan coordinates:
+            row1a_text = ""
+        else:
+            fmt = 'x={0:0' + str(maxsize) + '.1f} y={1:0' + str(maxsize) + '.1f}'
```
Very non-blocker comment!
Is there an opportunity here to use f-string?
```suggestion
            fmt = f'x={{0:0{maxsize}.1f}} y={{1:0{maxsize}.1f}}'
```

haticekaratay commented:
Thanks @bmorris3  for the detailed and clearly explained PR.
To test the performance difference, I used the Roman WFI calibration file (r0000101001001001001_0001_wfi01_cal.asdf). I compared the WCS from the ASDF file (gwcs.wcs._wcs.WCS) with a hand-built FITS WCS(similar to what’s being used in this PR)(astropy.wcs.WCS).
I ran both `pixel_to_world` and `world_to_pixel` for a 500×500 grid, repeating the tests 10 times and averaging the results.

Here are the results:
| Branch   | WCS Type             | `pixel_to_world` | `world_to_pixel` |
|----------|----------------------|------------------|------------------|
| This PR  | `astropy.wcs.WCS`    |    0.0389 s    |    0.0269 s  |
| Main     | `gwcs.wcs._wcs.WCS`  | 0.0480 s       | 0.0462 s      |

The FITS WCS used in the PR was consistently faster in both direction.

haticekaratay commented on jdaviz/configs/imviz/wcs_utils.py:
```diff
@@ -295,78 +292,34 @@ def data_outside_gwcs_bounding_box(data, x, y):
+        wcsaxes=2,
+        ctype1='RA---TAN',
+        ctype2='DEC--TAN',
+        crpix1=refdata_shape[0] / 2 + 0.75,
+        crpix2=refdata_shape[1] / 2 + 0.75,
```
Curious: Is there a known reason for the +0.75 ?

pllim commented:
@haticekaratay , just curious, on that same machine and env, how is the timing for 10k x 10k image?

https://github.com/spacetelescope/jdaviz/pull/3483#issuecomment-2755095408

haticekaratay commented:
> just curious, on that same machine and env, how is the timing for 10k x 10k image?

For 10k x 10k
WCS Type: <class 'astropy.wcs.wcs.WCS'>
pixel_to_world: mean = 15.2658s
world_to_pixel: mean = 15.2889s

WCS Type: <class 'gwcs.wcs._wcs.WCS'>
pixel_to_world: mean = 6.7963s
world_to_pixel: mean = 49.1957s

pllim commented:
Thanks, @haticekaratay for https://github.com/spacetelescope/jdaviz/pull/3483#issuecomment-2755191090 .

p.s. The asymmetric performance from GWCS is kinda surprising.

pllim commented on CHANGES.rst:
```diff
@@ -17,6 +17,8 @@ Cubeviz
 Imviz
 ^^^^^
 
+- Improve perfomance by using FITS WCS for reference data layers when linked by WCS, rather than GWCS [#3483]
```
```suggestion
- Improve perfomance by using FITS WCS for reference data layers when linked by WCS, rather than GWCS. [#3483]
```

pllim reviewed (commented):
I will review carefully tomorrow. A couple of comments now. Thanks!

bmorris3 commented:
@pllim RE https://github.com/spacetelescope/jdaviz/pull/3483#discussion_r2017507678: when WCS-linked, data with GWCS get called in both directions from pixel to world and world to pixel. If the world coordinates aren't defined, the pixel coordinates shouldn't be either.

pllim commented:
Re: https://github.com/spacetelescope/jdaviz/pull/3483#issuecomment-2761334855 -- Strictly speaking that change is unrelated to the performance goals and is taking a useful debugging feature away, so I feel a little sad seeing it go.

bmorris3 commented:
@pllim and I iterated on Slack and discovered that we didn't understand each other. The test the the comment above is acting correctly, but the coord info behavior regresses for single, unlinked images in Imviz. This is fixed in 7ee3255.

pllim reviewed (commented):
Should "Pixel" still show when cursor is out of bounds wrt GWCS and everything else is hidden?

![Screenshot 2025-03-31 145815](https://github.com/user-attachments/assets/41d80a1a-0db0-4d65-abc8-970214758bb0)

The example in `concepts/imviz_dithered_gwcs.ipynb` breaks down because now mouse is listening to top visible layer (B) so it refuses to show coordinates at (A) because it is following bounding box of B even though both are visible. I think this notebook reflects the original use case that made us ignore the bounding box. I guess we can open a follow-up ticket to revisit that, or just decide that this use case is no longer important to us. 🤷‍♀️ 

![Screenshot 2025-03-31 151117](https://github.com/user-attachments/assets/f5c85e94-9cef-4e77-8cdf-f27a38bbe950)

bmorris3 commented:
@pllim – I agree with your comment above, "Pixel" shouldn't show in that row in that case. I've adjusted the code accordingly and addressed your other comments.

pllim commented:
Thank you. The question about behavior change wrt `concepts/imviz_dithered_gwcs.ipynb` remains. I am not sure how Indigo wants to respond to that. Another dev will take over this review starting this coming sprint. Thank you for your patience.

bmorris3 commented:
@rosteen and I discussed the plan forward for coord info on mouseover at locations outside of the bounds of the selected/top image layer. We'll show world coords from the _reference data_ layer. This will be done in a follow-up PR as soon as this PR gets merged.

rosteen reviewed (approved):
Code looks good to me, I'm going to create a follow-up issue to fall back to the reference layer for WCS coordinates when the cursor is outside the bounding box of the top layer. Thanks!

rosteen commented:
I need to check to see if the last failure in https://github.com/spacetelescope/jdaviz/actions/runs/14383447257/job/40332784755?pr=3483 is known, I think the rest are all unrelated and known.
